util/tracecpu/profile_test.go

```diff
@@ -0,0 +1,253 @@
+// Copyright 2021 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package tracecpu_test
+
+import (
+	"bytes"
+	"context"
+	"crypto/sha256"
+	"encoding/hex"
+	"sync"
+	"testing"
+	"time"
+
+	"github.com/google/pprof/profile"
+	. "github.com/pingcap/check"
+	"github.com/pingcap/tidb/config"
+	"github.com/pingcap/tidb/util/hack"
+	"github.com/pingcap/tidb/util/tracecpu"
+	"github.com/uber-go/atomic"
+)
+
+func TestT(t *testing.T) {
+	CustomVerboseFlag = true
+	TestingT(t)
+}
+
+var _ = SerialSuites(&testSuite{})
+
+type testSuite struct{}
+
+func (s *testSuite) SetUpSuite(c *C) {
+	cfg := config.GetGlobalConfig()
+	newCfg := *cfg
+	newCfg.TopSQL.Enable = true
+	newCfg.TopSQL.RefreshInterval = 1
+	config.StoreGlobalConfig(&newCfg)
+	tracecpu.GlobalSQLStatsProfiler.Run()
+}
+
+func (s *testSuite) TestSQLStatsProfile(c *C) {
+	collector := newMockStatsCollector()
+	tracecpu.GlobalSQLStatsProfiler.SetCollector(collector)
+	reqs := []struct {
+		sql  string
+		plan string
+	}{
+		{"select * from t where a=?", "point-get"},
+		{"select * from t where a>?", "table-scan"},
+		{"insert into t values (?)", ""},
+	}
+	var wg sync.WaitGroup
+	for _, req := range reqs {
+		wg.Add(1)
+		go func(sql, plan string) {
+			defer wg.Done()
+			s.mockExecuteSQL(sql, plan)
+		}(req.sql, req.plan)
+	}
+
+	cnt := collector.getCollectCnt()
+	// test for StartCPUProfile.
+	buf := bytes.NewBuffer(nil)
+	err := tracecpu.StartCPUProfile(buf)
+	c.Assert(err, IsNil)
+	s.waitCollectCnt(collector, cnt+2)
```

Comment:
If the mock SQL execute goroutine finish rapidly before `StartCPUProfile` works, this test will be unstable?

Reply:
No, It is stable. The SQL execute goroutine doesn't affect the behavior of `tracecpu.StartCPUProfile`. The test is to make sure the `tracecpu.StartCPUProfile` API works fine and doesn't return an error.